File: springboot/src/main/java/com/mcp/server/tools/TemperatureTool.java
package com.mcp.server.tools;

import com.mcp.server.service.WeatherService;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Component;
import reactor.core.publisher.Mono;

import java.util.Map;

@Component
@RequiredArgsConstructor
public class TemperatureTool {

    private final WeatherService weatherService;

    public String getName() {
        return "get_temperature";
    }

    public String getDescription() {
        return "Get detailed temperature information for a specific city, including feels like, min, and max temperatures.";
    }

    public Map<String, Object> getInputSchema() {
        return Map.of(
            "type", "object",
            "properties", Map.of(
                "city", Map.of(
                    "type", "string",
                    "description", "Name of the city (e.g., 'Seoul', 'New York')"
                )
            ),
            "required", new String[]{"city"}
        );
    }

    public Mono<String> execute(Map<String, Object> arguments) {
        String city = (String) arguments.get("city");
        
        return weatherService.getWeatherData(city)
            .map(data -> {
                Map<String, Object> main = (Map<String, Object>) data.get("main");
                
                Object temp = main.get("temp");
                Object feelsLike = main.get("feels_like");
                Object tempMin = main.get("temp_min");
                Object tempMax = main.get("temp_max");
                
                return String.format("Temperature in %s:\n- Current: %s°C\n- Feels Like: %s°C\n- Min: %s°C\n- Max: %s°C",
                    city, temp, feelsLike, tempMin, tempMax);
            })
            .onErrorResume(e -> Mono.just("Error fetching temperature: " + e.getMessage()));
    }
}

File: python/src/tools/temperature_tool.py
from typing import Any, Dict
from src.services.weather_service import weather_service
from src.tool_registry import register_tool

@register_tool(
    name="get_temperature",
    description="Get detailed temperature information for a specific city, including feels like, min, and max temperatures.",
    parameters={
        "type": "object",
        "properties": {
            "city": {
                "type": "string",
                "description": "Name of the city (e.g., 'Seoul', 'New York')"
            }
        },
        "required": ["city"]
    }
)
async def get_temperature(city: str) -> str:
    try:
        data = await weather_service.get_weather_data(city)
        
        temp = data["main"]["temp"]
        feels_like = data["main"]["feels_like"]
        temp_min = data["main"]["temp_min"]
        temp_max = data["main"]["temp_max"]
        
        return (f"Temperature in {city}:\n"
                f"- Current: {temp}°C\n"
                f"- Feels Like: {feels_like}°C\n"
                f"- Min: {temp_min}°C\n"
                f"- Max: {temp_max}°C")
    except ValueError as e:
        return f"Error: {str(e)}"
    except Exception as e:
        return f"Error fetching temperature: {str(e)}"
